File: server/src/main/java/com/github/rodm/teamcity/gradle/scripts/server/DefaultGradleScriptsManager.java
/*
 * Copyright 2017 Rod MacKenzie.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.rodm.teamcity.gradle.scripts.server;

import jetbrains.buildServer.serverSide.CopiedObjects;
import jetbrains.buildServer.serverSide.CustomSettingsMapper;
import jetbrains.buildServer.serverSide.SProject;
import jetbrains.buildServer.util.ExceptionUtil;
import jetbrains.buildServer.util.FileUtil;
import org.apache.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.util.*;

import static jetbrains.buildServer.log.Loggers.SERVER_CATEGORY;
import static com.github.rodm.teamcity.gradle.scripts.GradleInitScriptsPlugin.PLUGIN_NAME;

public class DefaultGradleScriptsManager implements GradleScriptsManager, CustomSettingsMapper {

    @NotNull
    private static final Logger LOG = Logger.getLogger(SERVER_CATEGORY + ".GradleInitScripts");

    @Override
    public void mapData(@NotNull final CopiedObjects copiedObjects) {
        for (Map.Entry<SProject, SProject> entry: copiedObjects.getCopiedProjectsMap().entrySet()) {
            SProject source = entry.getKey();
            SProject target = entry.getValue();
            File sourceDir = source.getPluginDataDirectory(PLUGIN_NAME);
            File[] files = sourceDir.listFiles();
            if (files != null && files.length > 0) {
                File targetDir;
                try {
                    targetDir = FileUtil.createDir(target.getPluginDataDirectory(PLUGIN_NAME));
                } catch (IOException e) {
                    LOG.warn("Could not create directory for project Gradle init scripts", e);
                    continue;
                }
                for (File sourceFile : files) {
                    final File targetFile = new File(targetDir, sourceFile.getName());
                    try {
                        FileUtil.copy(sourceFile, targetFile);
                    } catch (IOException e) {
                        LOG.warn("Could not copy script file: " + sourceFile.getAbsolutePath() + " to: " + targetFile.getAbsolutePath(), e);
                    }
                }
            }
        }
    }

    @Override
    @NotNull
    public Map<SProject, List<String>> getScriptNames(@NotNull SProject project) {
        final Set<String> foundNames = new HashSet<>();
        final Map<SProject, List<String>> result = new LinkedHashMap<>();
        FileFilter filter = File::isFile;

        List<SProject> projectPath = project.getProjectPath();
        ListIterator<SProject> iter = projectPath.listIterator(projectPath.size());

        while (iter.hasPrevious()) {
            SProject currentProject = iter.previous();

            try {
                File pluginDataDirectory = getPluginDataDirectory(currentProject);
                File[] files = pluginDataDirectory.listFiles(filter);
                if (files != null && files.length > 0) {
                    List<String> scripts = new ArrayList<>();
                    for (File file : files) {
                        if (!foundNames.contains(file.getName())) {
                            scripts.add(file.getName());
                            foundNames.add(file.getName());
                        }
                    }
                    if (scripts.size() > 0) {
                        result.put(currentProject, scripts);
                    }
                }
            } catch (IOException e) {
                LOG.error(e.getMessage());
                result.clear();
            }
        }
        return result;
    }

    @Override
    public int getScriptsCount(@NotNull SProject project) {
        return getScriptNames(project).values().stream().mapToInt(List::size).sum();
    }

    @Override
    @Nullable
    public String findScript(@NotNull SProject project, @NotNull String name) {
        List<SProject> projectPath = project.getProjectPath();
        ListIterator<SProject> iter = projectPath.listIterator(projectPath.size());

        while (iter.hasPrevious()) {
            SProject currentProject = iter.previous();

            try {
                File file = new File(getPluginDataDirectory(currentProject), name);
                if (file.exists()) {
                    return FileUtil.readText(file, "UTF-8");
                }
            } catch (Exception e) {
                ExceptionUtil.rethrowAsRuntimeException(e);
            }
        }
        return null;
    }

    @Override
    public boolean deleteScript(@NotNull SProject project, @NotNull String name) {
        boolean result = false;
        try {
            File file = new File(getPluginDataDirectory(project), name);
            result = FileUtil.delete(file);
        }
        catch (IOException e) {
            LOG.error(e.getMessage());
        }
        return result;
    }

    @NotNull
    private File getPluginDataDirectory(@NotNull SProject project) throws IOException {
        return FileUtil.createDir(project.getPluginDataDirectory(PLUGIN_NAME));
    }
}

File: server/src/main/kotlin/com/github/rodm/teamcity/gradle/scripts/server/DefaultGradleScriptsManager.kt
/*
 * Copyright 2017 Rod MacKenzie.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.rodm.teamcity.gradle.scripts.server

import jetbrains.buildServer.controllers.FileSecurityUtil
import jetbrains.buildServer.log.Loggers.SERVER_CATEGORY
import jetbrains.buildServer.serverSide.ConfigActionFactory
import jetbrains.buildServer.serverSide.CopiedObjects
import jetbrains.buildServer.serverSide.CustomSettingsMapper
import jetbrains.buildServer.serverSide.SProject
import jetbrains.buildServer.serverSide.VersionedSettingsRegistry
import jetbrains.buildServer.serverSide.impl.ProjectEx
import jetbrains.buildServer.util.ExceptionUtil
import jetbrains.buildServer.util.FileUtil
import jetbrains.buildServer.web.openapi.PluginDescriptor
import org.apache.log4j.Logger
import java.io.File
import java.io.IOException
import kotlin.reflect.KFunction1

private const val DEFAULT_TIMEOUT = 10000L

class DefaultGradleScriptsManager(descriptor: PluginDescriptor,
                                  registry: VersionedSettingsRegistry,
                                  private val configActionFactory: ConfigActionFactory)
    : GradleScriptsManager, CustomSettingsMapper
{
    private val log = Logger.getLogger("$SERVER_CATEGORY.GradleInitScripts")

    private val pluginName = descriptor.pluginName

    init {
        registry.registerDir("pluginData/" + descriptor.pluginName)
    }

    override fun getScriptNames(project: SProject): Map<SProject, List<String>> {
        val foundNames = HashSet<String>()
        val result = LinkedHashMap<SProject, List<String>>()

        val projectPath = project.projectPath
        val iter = projectPath.listIterator(projectPath.size)

        while (iter.hasPrevious()) {
            val currentProject = iter.previous()

            try {
                val scripts = getScriptNamesForProject(currentProject)
                scripts.removeAll(foundNames)
                foundNames.addAll(scripts)
                if (scripts.isNotEmpty()) {
                    result[currentProject] = scripts
                }
            } catch (e: IOException) {
                log.error(e.message)
                result.clear()
            }
        }
        return result
    }

    private val filter: KFunction1<File, Boolean> = File::isFile

    private fun getScriptNamesForProject(project: SProject): MutableList<String> {
        val scripts = mutableListOf<String>()
        val files = getPluginDataDirectory(project).listFiles(filter)
        if (files != null && files.isNotEmpty()) {
            for (file in files) {
                scripts.add(file.name)
            }
        }
        return scripts
    }

    override fun getScriptsCount(project: SProject): Int {
        return getScriptNames(project).values.stream().mapToInt { it.size }.sum()
    }

    override fun findScript(project: SProject, name: String): String? {
        val projectPath = project.projectPath
        val iter = projectPath.listIterator(projectPath.size)

        while (iter.hasPrevious()) {
            val currentProject = iter.previous()

            try {
                val file = File(getPluginDataDirectory(currentProject), name)
                if (file.exists()) {
                    return FileUtil.readText(file, "UTF-8")
                }
            } catch (e: Exception) {
                ExceptionUtil.rethrowAsRuntimeException(e)
            }
        }
        return null
    }

    override fun saveScript(project: SProject, name: String, content: String): Boolean {
        val file = File(getPluginDataDirectory(project), name)
        val exists = file.exists()
        val message = "Gradle init script $name was ${if (exists) "updated" else "uploaded"}"
        val action = configActionFactory.createAction(project, message)
        val projectEx = project as ProjectEx
        val task = projectEx.scheduleFileSave(action, name, content.toByteArray())
        return try {
            task.await(DEFAULT_TIMEOUT)
        }
        catch (_: InterruptedException) {
            false
        }
    }

    override fun deleteScript(project: SProject, name: String): Boolean {
        val message = "Gradle init script $name was deleted"
        val projectEx = project as ProjectEx
        val task = projectEx.scheduleFileDelete(configActionFactory.createAction(project, message), name)
        return try {
            task.await(DEFAULT_TIMEOUT)
        }
        catch (_: InterruptedException) {
            false
        }
    }

    override fun mapData(copiedObjects: CopiedObjects) {
        for ((source, target) in copiedObjects.copiedProjectsMap) {
            val sourceDir = source.getPluginDataDirectory(pluginName)
            val files = sourceDir.listFiles()
            if (files != null && files.isNotEmpty()) {
                for (sourceFile in files) {
                    val relPath = getValidRelativePath(source, sourceFile.name)
                    try {
                        val action = configActionFactory.createAction("Copy Gradle init script")
                        (target as ProjectEx).scheduleFileSave(action, relPath, FileUtil.loadFileBytes(sourceFile))
                    } catch (e: IOException) {
                        log.warn("Could not load contents of init script file: " + sourceFile.absolutePath, e)
                    }
                }
            }
        }
    }

    fun getValidRelativePath(project: SProject, name: String): String {
        val pluginDataDir = getPluginDataDirectory(project)
        val scriptPath = pluginDataDir.toPath().resolve(name)
        val configDir = project.configDirectory.toPath()
        FileSecurityUtil.checkInsideDirectory(scriptPath.toFile(), pluginDataDir)
        return configDir.relativize(scriptPath).toString()
    }

    private fun getPluginDataDirectory(project: SProject): File {
        return FileUtil.createDir(project.getPluginDataDirectory(pluginName))
    }
}
